src/schedule/zephyr_dp_schedule.c

```diff
@@ -34,20 +37,59 @@ struct scheduler_dp_data {
 
 struct task_dp_pdata {
 	k_tid_t thread_id;		/* zephyr thread ID */
-	struct k_thread thread;		/* memory space for a thread */
+	struct k_thread *thread;	/* pointer to the kernels' thread object */
+	struct k_thread thread_struct;	/* thread object for kernel threads */
 	uint32_t deadline_clock_ticks;	/* dp module deadline in Zephyr ticks */
 	k_thread_stack_t __sparse_cache *p_stack;	/* pointer to thread stack */
 	size_t stack_size;		/* size of the stack in bytes */
-	struct k_sem sem;		/* semaphore for task scheduling */
+	struct k_sem *sem;		/* pointer to semaphore for task scheduling */
+	struct k_sem sem_struct;	/* semaphore for task scheduling for kernel threads */
 	struct processing_module *mod;	/* the module to be scheduled */
 	uint32_t ll_cycles_to_start;    /* current number of LL cycles till delayed start */
 };
 
+#ifdef CONFIG_USERSPACE
+/* Single CPU-wide lock
+ * The irq_lock is not available for USERSPACE (non-privileged) threads.
+ * Therefore semaphore is used to control critical section.
+ */
+#define DP_LOCK_INIT(i, _)	Z_SEM_INITIALIZER(dp_lock[i], 1, 1)
+#define DP_LOCK_INIT_LIST	LISTIFY(CONFIG_MP_MAX_NUM_CPUS, DP_LOCK_INIT, (,))
+
```

Comment:
We can revisit this in a subsequent PR to make this a more global locking API as there will be other users.